File: template/modules/react-native-vision-engine/ios/HybridVisionEngine.swift
import AVFoundation
import Foundation
import NitroModules
import VisionCamera

/// Фасад nitro-спеки `VisionEngine`: держит слоты моделей и оркеструет
/// сервисы (`VisionTextRecognizer`, `CoreMLObjectDetector`,
/// `CoreMLModelLoader`, `FrameGeometry`). Логики распознавания здесь нет.
/// Все методы обработки синхронные — вызываются с frame-потока VisionCamera.
///
/// Слоты моделей — per-instance (у каждого сканера свой движок), сами
/// модели кэшируются `CoreMLModelLoader` по имени на всё приложение.
class HybridVisionEngine: HybridVisionEngineSpec {
  /// Фолбэки порогов детектора для вызовов без соответствующих полей опций.
  /// Обязаны совпадать с `DETECTOR_DEFAULTS` JS-модуля (рантайм-источник —
  /// значения, переданные в опциях).
  private enum DetectorDefaults {
    static let regionMinScore = 0.35
    static let maxRegions = 6.0
    static let maxRegionsPerClass = 2.0
    static let regionPadding = 0.18
    static let iouThreshold = 0.45
  }

  /// Запись слота — из async-контекста загрузки, чтение — с frame-потока;
  /// доступ только под `modelLock`. lock/unlock недоступны из async-контекста
  /// (Swift 6) — запись инкапсулирована в синхронные методы `store*`.
  private let modelLock = NSLock()
  private var detectorSlot: LoadedCoreMLModel?
  private var objectModelSlot: LoadedCoreMLModel?

  private var detector: LoadedCoreMLModel? {
    modelLock.lock()
    defer { modelLock.unlock() }
    return detectorSlot
  }

  private var objectModel: LoadedCoreMLModel? {
    modelLock.lock()
    defer { modelLock.unlock() }
    return objectModelSlot
  }

  private func storeDetector(_ model: LoadedCoreMLModel) {
    modelLock.lock()
    detectorSlot = model
    modelLock.unlock()
  }

  private func storeObjectModel(_ model: LoadedCoreMLModel) {
    modelLock.lock()
    objectModelSlot = model
    modelLock.unlock()
  }

  var isDetectorLoaded: Bool {
    return detector != nil
  }

  var isObjectModelLoaded: Bool {
    return objectModel != nil
  }

  func loadDetector(modelName: String) throws -> Promise<Bool> {
    return Promise.async {
      guard let loaded = try await CoreMLModelLoader.load(named: modelName) else {
        return false
      }
      self.storeDetector(loaded)
      return true
    }
  }

  func loadObjectModel(modelName: String) throws -> Promise<Bool> {
    return Promise.async {
      guard let loaded = try await CoreMLModelLoader.load(named: modelName) else {
        return false
      }
      self.storeObjectModel(loaded)
      return true
    }
  }

  // MARK: - Spec methods

  func scan(frame: any HybridFrameSpec, options: OcrScanOptions) throws -> OcrScanResult {
    return try runOcr(Self.unwrap(frame), options: options)
  }

  func detectObjects(frame: any HybridFrameSpec, options: ObjectScanOptions) throws -> ObjectScanResult {
    return try runObjects(Self.unwrap(frame), options: options)
  }

  func analyze(frame: any HybridFrameSpec, options: AnalyzeOptions) throws -> AnalyzeResult {
    let context = try Self.unwrap(frame)
    var ocr: OcrScanResult?
    var objects: ObjectScanResult?

    if let ocrOptions = options.ocr {
      ocr = try runOcr(context, options: ocrOptions)
    }
    if let objectOptions = options.objects {
      objects = try runObjects(context, options: objectOptions)
    }

    return AnalyzeResult(ocr: ocr, objects: objects)
  }

  // MARK: - Pipelines

  /// OCR кадра: с детектором — только по его регионам (ROI + padding,
  /// батч запросов одним handler'ом), без детектора или при пустых
  /// регионах — полнокадрово
  private func runOcr(
    _ context: FrameContext,
    options: OcrScanOptions
  ) throws -> OcrScanResult {
    let start = CACurrentMediaTime()
    let detector = self.detector

    var regions: [RawDetection] = []
    var observations: [OcrObservation] = []
    if let detector {
      // детектор наводит OCR на регионы интереса — читаем только их
      let detections = try CoreMLObjectDetector.detect(
        detector,
        pixelBuffer: context.pixelBuffer,
        orientation: context.orientation,
        minScore: Float(options.regionMinScore ?? DetectorDefaults.regionMinScore),
        iouThreshold: CGFloat(options.regionIouThreshold ?? DetectorDefaults.iouThreshold)
      )
      regions = Self.selectRegions(detections, options: options)
      if !regions.isEmpty {
        let padding = CGFloat(options.regionPadding ?? DetectorDefaults.regionPadding)
        let rois = regions.map { detection in
          OcrRegionOfInterest(
            rect: FrameGeometry.pad(FrameGeometry.toVisionROI(detection.rect), by: padding),
            classIndex: detection.classIndex
          )
        }
        observations = try VisionTextRecognizer.recognize(
          in: context.pixelBuffer,
          orientation: context.orientation,
          regions: rois,
          options: options
        )
      }
    }
    // без детектора полный кадр — единственный режим; с детектором полный
    // кадр читается только при явном fullFrameFallback и пустых кропах
    let fallbackAllowed = options.fullFrameFallback ?? false
    if detector == nil || (observations.isEmpty && fallbackAllowed) {
      observations = try VisionTextRecognizer.recognize(
        in: context.pixelBuffer,
        orientation: context.orientation,
        regions: nil,
        options: options
      )
    }

    observations = Self.limitObservations(
      observations
        .filter { $0.confidence >= options.minConfidence }
        .sorted { $0.confidence > $1.confidence },
      limit: max(0, Int(options.maxObservations))
    )

    return OcrScanResult(
      observations: observations,
      regions: regions.map(Self.toDetectedObject),
      // Vision возвращает боксы уже в системе координат ориентированного
      // (выпрямленного) изображения — доворот на JS-стороне не нужен
      bufferOrientation: .up,
      imageWidth: context.uprightWidth,
      imageHeight: context.uprightHeight,
      durationMs: (CACurrentMediaTime() - start) * 1000,
      detectorUsed: detector != nil
    )
  }

  /// Детекция объектов кадра моделью из слота `objectModel`
  private func runObjects(
    _ context: FrameContext,
    options: ObjectScanOptions
  ) throws -> ObjectScanResult {
    let start = CACurrentMediaTime()
    guard let objectModel = self.objectModel else {
      throw RuntimeError.error(withMessage: "VisionEngine: object model is not loaded — call loadObjectModel() first")
    }

    var detections = try CoreMLObjectDetector.detect(
      objectModel,
      pixelBuffer: context.pixelBuffer,
      orientation: context.orientation,
      minScore: Float(options.minScore),
      iouThreshold: CGFloat(options.iouThreshold ?? DetectorDefaults.iouThreshold)
    )
    let limit = max(0, Int(options.maxObjects))
    if detections.count > limit {
      detections = Array(detections.prefix(limit))
    }

    return ObjectScanResult(
      objects: detections.map(Self.toDetectedObject),
      bufferOrientation: .up,
      imageWidth: context.uprightWidth,
      imageHeight: context.uprightHeight,
      durationMs: (CACurrentMediaTime() - start) * 1000
    )
  }

  /// Обрезка до лимита по кругу между регионами: строк таблички кратно
  /// больше, чем у номера, и глобальный top-N по уверенности вытеснил бы
  /// малые регионы целиком. Внутри региона порядок по уверенности сохраняется.
  private static func limitObservations(
    _ observations: [OcrObservation],
    limit: Int
  ) -> [OcrObservation] {
    if observations.count <= limit {
      return observations
    }

    var order: [Double] = []
    var groups: [Double: [OcrObservation]] = [:]
    for observation in observations {
      let key = observation.regionClassIndex
      if groups[key] == nil {
        order.append(key)
        groups[key] = []
      }
      groups[key]?.append(observation)
    }

    var result: [OcrObservation] = []
    var index = 0
    while result.count < limit {
      var appended = false
      for key in order {
        guard let group = groups[key], index < group.count else {
          continue
        }
        result.append(group[index])
        appended = true
        if result.count >= limit {
          break
        }
      }
      if !appended {
        break
      }
      index += 1
    }

    return result
  }

  /// Регионы детектора → те, что уходят в OCR: фильтр по разрешённым
  /// классам, затем квота на класс и общий лимит. Детекции приходят
  /// отсортированными по score, порядок сохраняется.
  private static func selectRegions(
    _ detections: [RawDetection],
    options: OcrScanOptions
  ) -> [RawDetection] {
    let maxRegions = max(0, Int(options.maxRegions ?? DetectorDefaults.maxRegions))
    let perClassLimit = Int(options.maxRegionsPerClass ?? DetectorDefaults.maxRegionsPerClass)
    let maxPerClass = perClassLimit > 0 ? perClassLimit : Int.max
    let allowed = options.regionClasses.flatMap { classes -> Set<Int32>? in
      classes.isEmpty ? nil : Set(classes.map { Int32($0) })
    }

    var counts: [Int32: Int] = [:]
    var selected: [RawDetection] = []
    for detection in detections {
      if selected.count >= maxRegions {
        break
      }
      if let allowed, !allowed.contains(detection.classIndex) {
        continue
      }
      let used = counts[detection.classIndex] ?? 0
      if used >= maxPerClass {
        continue
      }
      counts[detection.classIndex] = used + 1
      selected.append(detection)
    }

    return selected
  }

  // MARK: - Frame plumbing

  /// Распакованный кадр: pixel buffer, EXIF-ориентация для Vision и
  /// размеры выпрямленного изображения. Общий для всех конвейеров вызова.
  private struct FrameContext {
    let pixelBuffer: CVPixelBuffer
    let orientation: CGImagePropertyOrientation
    let uprightWidth: Double
    let uprightHeight: Double
  }

  /// Frame VisionCamera → контекст обработки; бросает, если кадр освобождён
  private static func unwrap(_ frame: any HybridFrameSpec) throws -> FrameContext {
    guard let nativeFrame = frame as? NativeFrame,
          let pixelBuffer = nativeFrame.sampleBuffer?.imageBuffer else {
      throw RuntimeError.error(withMessage: "VisionEngine: Frame has no pixel buffer — is it already disposed?")
    }

    let orientation = FrameGeometry.cgOrientation(
      frame.orientation,
      isMirrored: frame.isMirrored
    )
    let bufferWidth = CVPixelBufferGetWidth(pixelBuffer)
    let bufferHeight = CVPixelBufferGetHeight(pixelBuffer)
    let isRotated = orientation.swapsDimensions

    return FrameContext(
      pixelBuffer: pixelBuffer,
      orientation: orientation,
      uprightWidth: Double(isRotated ? bufferHeight : bufferWidth),
      uprightHeight: Double(isRotated ? bufferWidth : bufferHeight)
    )
  }

  /// Внутренняя детекция → `DetectedObject` контракта спеки
  private static func toDetectedObject(_ detection: RawDetection) -> DetectedObject {
    return DetectedObject(
      classIndex: Double(detection.classIndex),
      label: detection.label,
      score: Double(detection.score),
      rect: OcrRect(
        x: detection.rect.minX,
        y: detection.rect.minY,
        width: detection.rect.width,
        height: detection.rect.height
      )
    )
  }
}

File: template/modules/react-native-vision-engine/android/src/main/java/com/margelo/nitro/visionengine/HybridVisionEngine.kt
package com.margelo.nitro.visionengine

import android.graphics.Bitmap
import android.os.SystemClock
import androidx.camera.core.ExperimentalGetImage
import androidx.camera.core.ImageProxy
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.camera.public.NativeFrame
import com.margelo.nitro.core.Promise
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * Фасад nitro-спеки `VisionEngine`: держит слоты моделей и оркеструет
 * сервисы (`MlKitTextRecognizer`, `TfliteDetector`, `FrameGeometry`).
 * Логики распознавания здесь нет. Все методы обработки синхронные —
 * вызываются с frame-потока VisionCamera (не main).
 *
 * Слоты моделей — per-instance (у каждого сканера свой движок); сами
 * `TfliteDetector` кэшируются по имени на всё приложение, поэтому слот
 * не владеет моделью и не закрывает её. Запись слота — из async-контекста
 * загрузки, чтение — с frame-потока (`@Volatile`).
 */
@DoNotStrip
class HybridVisionEngine : HybridVisionEngineSpec() {
  /** Детектор регионов для наведения OCR (слот `loadDetector`) */
  @Volatile
  private var detector: TfliteDetector? = null
  /** Модель детекции объектов (слот `loadObjectModel`) */
  @Volatile
  private var objectDetector: TfliteDetector? = null

  override val isDetectorLoaded: Boolean
    get() = detector != null

  override val isObjectModelLoaded: Boolean
    get() = objectDetector != null

  override fun loadDetector(modelName: String): Promise<Boolean> {
    return Promise.async {
      val context = NitroModules.applicationContext ?: return@async false
      val loaded = TfliteDetector.load(context, "$modelName.tflite") ?: return@async false
      detector = loaded
      true
    }
  }

  override fun loadObjectModel(modelName: String): Promise<Boolean> {
    return Promise.async {
      val context = NitroModules.applicationContext ?: return@async false
      val loaded = TfliteDetector.load(context, "$modelName.tflite") ?: return@async false
      objectDetector = loaded
      true
    }
  }

  // ─── Методы спеки ──────────────────────────────────────────────────────────

  override fun scan(frame: HybridFrameSpec, options: OcrScanOptions): OcrScanResult {
    return withFrame(frame) { context -> runOcr(context, options) }
  }

  override fun detectObjects(
    frame: HybridFrameSpec,
    options: ObjectScanOptions,
  ): ObjectScanResult {
    return withFrame(frame) { context -> runObjects(context, options) }
  }

  override fun analyze(frame: HybridFrameSpec, options: AnalyzeOptions): AnalyzeResult {
    return withFrame(frame) { context ->
      AnalyzeResult(
        ocr = options.ocr?.let { runOcr(context, it) },
        objects = options.objects?.let { runObjects(context, it) },
      )
    }
  }

  // ─── Конвейеры ─────────────────────────────────────────────────────────────

  /**
   * OCR кадра: с детектором — только по кропам его регионов, без детектора
   * или при пустых регионах — полный кадр через ML Kit.
   */
  private fun runOcr(context: FrameContext, options: OcrScanOptions): OcrScanResult {
    val start = SystemClock.elapsedRealtime()

    val detector = this.detector
    var regions: List<DetectedRegion> = emptyList()
    var observations: List<OcrObservation> = emptyList()
    if (detector != null) {
      // детектор наводит OCR на регионы интереса — читаем только кропы
      val upright = context.upright()
      val regionPadding = (options.regionPadding ?: DEFAULT_REGION_PADDING).toFloat()

      regions = selectRegions(
        detector.detect(
          upright,
          minScore = (options.regionMinScore ?: DEFAULT_REGION_MIN_SCORE).toFloat(),
          iouThreshold = (options.regionIouThreshold ?: DEFAULT_NMS_IOU).toFloat(),
        ),
        options,
      )
      observations = regions.flatMap { region ->
        val crop = FrameGeometry.cropRegion(
          upright,
          region,
          padding = regionPadding,
          minSizePx = MIN_REGION_SIZE_PX,
        ) ?: return@flatMap emptyList()
        try {
          val text = MlKitTextRecognizer.recognize(crop.bitmap)
          MlKitTextRecognizer
            .toObservations(
              text,
              crop.bitmap.width,
              crop.bitmap.height,
              regionClassIndex = region.classIndex,
            )
            .map { observation ->
              observation.copy(
                rect = FrameGeometry.shiftRect(
                  observation.rect,
                  crop,
                  context.imageWidth,
                  context.imageHeight,
                ),
              )
            }
        } finally {
          crop.bitmap.recycle()
        }
      }
    }
    // без детектора полный кадр — единственный режим; с детектором полный
    // кадр читается только при явном fullFrameFallback и пустых кропах
    val fallbackAllowed = options.fullFrameFallback ?: false
    if (detector == null || (observations.isEmpty() && fallbackAllowed)) {
      val mediaImage = context.proxy.image
        ?: throw Error("VisionEngine: Frame has no media Image — is it already disposed?")
      val text = MlKitTextRecognizer.recognize(mediaImage, context.rotation)
      observations = MlKitTextRecognizer.toObservations(
        text,
        context.imageWidth,
        context.imageHeight,
        regionClassIndex = MlKitTextRecognizer.FULL_FRAME_CLASS_INDEX,
      )
    }

    val filtered = limitObservations(
      observations
        .filter { it.confidence >= options.minConfidence }
        .sortedByDescending { it.confidence },
      max(0, options.maxObservations.roundToInt()),
    )

    return OcrScanResult(
      observations = filtered.toTypedArray(),
      regions = regions.map(::toDetectedObject).toTypedArray(),
      // ML Kit получает rotationDegrees и отдаёт боксы уже в выпрямленных
      // координатах — дополнительное преобразование не требуется
      bufferOrientation = OcrBufferOrientation.UP,
      imageWidth = context.imageWidth.toDouble(),
      imageHeight = context.imageHeight.toDouble(),
      durationMs = (SystemClock.elapsedRealtime() - start).toDouble(),
      detectorUsed = detector != null,
    )
  }

  /** Детекция объектов кадра моделью из слота `objectDetector` */
  private fun runObjects(context: FrameContext, options: ObjectScanOptions): ObjectScanResult {
    val start = SystemClock.elapsedRealtime()
    val objectDetector = this.objectDetector
      ?: throw Error("VisionEngine: object model is not loaded — call loadObjectModel() first")

    val objects = objectDetector
      .detect(
        context.upright(),
        minScore = options.minScore.toFloat(),
        iouThreshold = (options.iouThreshold ?: DEFAULT_NMS_IOU).toFloat(),
      )
      .take(max(0, options.maxObjects.roundToInt()))
      .map(::toDetectedObject)

    return ObjectScanResult(
      objects = objects.toTypedArray(),
      // детекция идёт по выпрямленному bitmap — координаты уже upright
      bufferOrientation = OcrBufferOrientation.UP,
      imageWidth = context.imageWidth.toDouble(),
      imageHeight = context.imageHeight.toDouble(),
      durationMs = (SystemClock.elapsedRealtime() - start).toDouble(),
    )
  }

  /**
   * Обрезка до лимита по кругу между регионами: строк таблички кратно больше,
   * чем у номера, и глобальный top-N по уверенности вытеснил бы малые регионы
   * целиком. Внутри региона порядок по уверенности сохраняется.
   */
  private fun limitObservations(
    observations: List<OcrObservation>,
    limit: Int,
  ): List<OcrObservation> {
    if (observations.size <= limit) {
      return observations
    }

    val groups = LinkedHashMap<Double, MutableList<OcrObservation>>()
    for (observation in observations) {
      groups.getOrPut(observation.regionClassIndex) { ArrayList() }.add(observation)
    }

    val result = ArrayList<OcrObservation>(limit)
    var index = 0
    while (result.size < limit) {
      var appended = false
      for (group in groups.values) {
        if (index >= group.size) {
          continue
        }
        result.add(group[index])
        appended = true
        if (result.size >= limit) {
          break
        }
      }
      if (!appended) {
        break
      }
      index++
    }

    return result
  }

  /**
   * Регионы детектора → те, что уходят в OCR: фильтр по разрешённым классам,
   * затем квота на класс и общий лимит. Детекции приходят отсортированными
   * по score, порядок сохраняется.
   */
  private fun selectRegions(
    detections: List<DetectedRegion>,
    options: OcrScanOptions,
  ): List<DetectedRegion> {
    val maxRegions = max(0, (options.maxRegions ?: DEFAULT_MAX_REGIONS).roundToInt())
    val perClassLimit = (options.maxRegionsPerClass ?: DEFAULT_MAX_REGIONS_PER_CLASS).roundToInt()
    val maxPerClass = if (perClassLimit > 0) perClassLimit else Int.MAX_VALUE
    val allowed = options.regionClasses
      ?.takeIf { it.isNotEmpty() }
      ?.map { it.roundToInt() }
      ?.toSet()

    val counts = HashMap<Int, Int>()
    val selected = ArrayList<DetectedRegion>(maxRegions)
    for (detection in detections) {
      if (selected.size >= maxRegions) {
        break
      }
      if (allowed != null && detection.classIndex !in allowed) {
        continue
      }
      val used = counts[detection.classIndex] ?: 0
      if (used >= maxPerClass) {
        continue
      }
      counts[detection.classIndex] = used + 1
      selected.add(detection)
    }

    return selected
  }

  // ─── Кадр ──────────────────────────────────────────────────────────────────

  /**
   * Контекст одного кадра: выпрямленный bitmap считается лениво и один раз
   * на все конвейеры вызова (`analyze` делит его между OCR и объектами).
   */
  private class FrameContext(val proxy: ImageProxy, val rotation: Int) {
    val imageWidth: Int =
      if (rotation == 90 || rotation == 270) proxy.height else proxy.width
    val imageHeight: Int =
      if (rotation == 90 || rotation == 270) proxy.width else proxy.height

    private var uprightBitmap: Bitmap? = null

    fun upright(): Bitmap {
      return uprightBitmap
        ?: FrameGeometry.uprightBitmap(proxy, rotation).also { uprightBitmap = it }
    }

    fun release() {
      uprightBitmap?.recycle()
      uprightBitmap = null
    }
  }

  /** Frame VisionCamera → контекст обработки; upright-битмап освобождается после `block` */
  @OptIn(ExperimentalGetImage::class)
  private inline fun <T> withFrame(frame: HybridFrameSpec, block: (FrameContext) -> T): T {
    val nativeFrame = frame as? NativeFrame
      ?: throw Error("VisionEngine: unexpected Frame implementation — expected VisionCamera NativeFrame")
    val context = FrameContext(nativeFrame.image, nativeFrame.image.imageInfo.rotationDegrees)
    try {
      return block(context)
    } finally {
      context.release()
    }
  }

  /** Внутренняя детекция → `DetectedObject` контракта спеки (меток у TFLite нет) */
  private fun toDetectedObject(region: DetectedRegion): DetectedObject {
    return DetectedObject(
      classIndex = region.classIndex.toDouble(),
      label = "",
      score = region.score.toDouble(),
      rect = OcrRect(
        x = region.x.toDouble(),
        y = region.y.toDouble(),
        width = region.width.toDouble(),
        height = region.height.toDouble(),
      ),
    )
  }

  private companion object {
    // Фолбэки порогов детектора для вызовов без соответствующих полей
    // опций; обязаны совпадать с `DETECTOR_DEFAULTS` JS-модуля
    // (рантайм-источник — значения, переданные в опциях)
    const val DEFAULT_REGION_MIN_SCORE = 0.35
    const val DEFAULT_MAX_REGIONS = 6.0
    const val DEFAULT_MAX_REGIONS_PER_CLASS = 2.0
    const val DEFAULT_REGION_PADDING = 0.18
    const val DEFAULT_NMS_IOU = 0.45

    /** Минимальный размер кропа, при котором OCR имеет смысл */
    const val MIN_REGION_SIZE_PX = 32
  }
}
